[ARMv7] Fix WasmAddress testb3

justinmichaud opened the pull request:
#### 7f25cfeaf6f216f0526604f90fe9b2bec74253ec
<pre>
[ARMv7] Fix WasmAddress testb3
<a href="https://bugs.webkit.org/show_bug.cgi?id=283938">https://bugs.webkit.org/show_bug.cgi?id=283938</a>

Reviewed by Yusuke Suzuki.

For now, WasmAddress is only being used by testb3. In the future,
we may want to pin a register on armv7 too, and it was really easy
to make this work, so we just went ahead and did it.

We also comment out some testb3 cases on armv7 using Check, which we aren&apos;t
planning to support (until it gets used in WASM).

* Source/JavaScriptCore/b3/B3LowerInt64.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
* Source/JavaScriptCore/b3/B3WasmAddressValue.cpp:
(JSC::B3::WasmAddressValue::WasmAddressValue):
* Source/JavaScriptCore/b3/air/AirCustom.cpp:
(JSC::B3::Air::WasmBoundsCheckCustom::generate):
* Source/JavaScriptCore/b3/air/AirCustom.h:
(JSC::B3::Air::WasmBoundsCheckCustom::forEachArg):
* Source/JavaScriptCore/b3/testb3_1.cpp:
(run):
* Source/JavaScriptCore/b3/testb3_3.cpp:
(addShrTests):
* Source/JavaScriptCore/b3/testb3_6.cpp:
(testLateRegister):
* Source/JavaScriptCore/b3/testb3_7.cpp:
(testPinRegisters):
* Source/JavaScriptCore/b3/testb3_8.cpp:
(testWasmAddressDoesNotCSE):
(testStoreAfterClobberDifferentWidth):
(testStoreAfterClobberDifferentWidthSuccessor):
(testStoreAfterClobberExitsSidewaysSuccessor):
* Source/JavaScriptCore/b3/testb3_failingArmV7Tests.inc:

Canonical link: <a href="https://commits.webkit.org/288153@main">https://commits.webkit.org/288153@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d17f29d39de3bf37b9eae0ebbacd50bec9d0d8d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81857 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1383 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35811 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86411 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32853 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1416 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9205 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63843 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21573 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84927 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1043 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74491 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44129 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/942 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28666 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31306 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74837 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72297 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29275 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87841 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80911 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9097 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6466 "Found 1 new test failure: swipe/navigate-event-back-swipe-verify-ua-transition.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72197 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9282 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70312 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71426 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17834 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15521 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14439 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/472 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9048 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14580 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103323 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8889 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25087 "Found 2 new JSC stress test failures: stress/json-stringify-inspector-check.js.no-llint, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12412 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10697 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aoikonomopoulos commented on Source/JavaScriptCore/b3/testb3_failingArmV7Tests.inc:
```diff
@@ -1,40 +1,4 @@
 "tailDupedTuplePair",
-"testCall",
-"testCallFunctionWithHellaArguments",
-"testCallFunctionWithHellaArguments3",
```
Nit, I think this is still failing after this patch, probably need some variant of https://github.com/WebKit/WebKit/pull/36092.

Constellation commented on Source/JavaScriptCore/jit/Width.h:
```diff
@@ -44,6 +44,12 @@ static constexpr Width Width32 = Width::Width32;
 static constexpr Width Width64 = Width::Width64;
 static constexpr Width Width128 = Width::Width128;
 
+#if USE(JSVALUE64)
```
It should be `CPU(ADDRESS64)`

justinmichaud committed:
[ARMv7] Fix WasmAddress testb3
https://bugs.webkit.org/show_bug.cgi?id=283938

Reviewed by Yusuke Suzuki.

For now, WasmAddress is only being used by testb3. In the future,
we may want to pin a register on armv7 too, and it was really easy
to make this work, so we just went ahead and did it.

We also comment out some testb3 cases on armv7 using Check, which we aren't
planning to support (until it gets used in WASM).

* Source/JavaScriptCore/b3/B3LowerInt64.cpp:
* Source/JavaScriptCore/b3/B3LowerToAir32_64.cpp:
* Source/JavaScriptCore/b3/B3WasmAddressValue.cpp:
(JSC::B3::WasmAddressValue::WasmAddressValue):
* Source/JavaScriptCore/b3/air/AirCustom.cpp:
(JSC::B3::Air::WasmBoundsCheckCustom::generate):
* Source/JavaScriptCore/b3/air/AirCustom.h:
(JSC::B3::Air::WasmBoundsCheckCustom::forEachArg):
* Source/JavaScriptCore/b3/testb3_1.cpp:
(run):
* Source/JavaScriptCore/b3/testb3_3.cpp:
(addShrTests):
* Source/JavaScriptCore/b3/testb3_6.cpp:
(testLateRegister):
* Source/JavaScriptCore/b3/testb3_7.cpp:
(testPinRegisters):
* Source/JavaScriptCore/b3/testb3_8.cpp:
(testWasmAddressDoesNotCSE):
(testStoreAfterClobberDifferentWidth):
(testStoreAfterClobberDifferentWidthSuccessor):
(testStoreAfterClobberExitsSidewaysSuccessor):
* Source/JavaScriptCore/b3/testb3_failingArmV7Tests.inc:

Canonical link: https://commits.webkit.org/288153@main
